cli/config/validator.go

```diff
@@ -0,0 +1,105 @@
+package config
+
+import (
+	"cmp"
+	"slices"
+)
+
+type pxfDeploymentValidator func(d *PxfDeployment) error
+type pxfClusterValidator func(c *PxfCluster) error
+type pxfServiceGroupValidator func(c *PxfCluster, sg *PxfServiceGroup) error
+
+
+// Validate applies deployment-level validation rules
+// and then calls validation for every cluster in the deployment
+func (d *PxfDeployment) Validate() error {
+	var validators = []pxfDeploymentValidator{
+		pxfDeploymentDefaultCluster,
+		pxfDeploymentHostnameUnique,
+		pxfDeploymentNoMoreThanOneCollocated,
+		pxfDeploymentGroupNameUnique,
+	}
+	// apply deployment-level validators
+	for _, validator := range validators {
+		if err := validator(d); err != nil {
+			return err
+		}
+	}
+	// validate clusters iterating in the order of their names
+	for _, cluster := range sortedMapValues(d.Clusters) {
+		if err := cluster.validate(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// validate applies cluster-level validation rules
+// and then calls validation for every service group and host of the cluster
+func (c *PxfCluster) validate() error {
+	var validators = []pxfClusterValidator{
+		pxfClusterIgnoreHostsWhenCollocatedWarning,
+		pxfClusterIgnoreEndpointWhenCollocatedWarning,
+		pxfClusterEndpointOrHostsMustExistWhenExternal,
+		pxfClusterValidEndpoint,
+		pxfClusterAtLeastOneServiceGroup,
+	}
+	// apply cluster-level validators
+	for _, validator := range validators {
+		if err := validator(c); err != nil {
+			return err
+		}
+	}
+	// validate service groups iterating in the order of their names
+	for _, group := range sortedMapValues(c.Groups) {
+		if err := group.validate(c); err != nil {
+			return err
+		}
+	}
+	// validate hosts
+	for _, host := range c.Hosts {
+		if err := host.validate(c); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// validate applies cluster-level validation rules
+// a pointer to a cluster needs to be passed to give proper context for the error message
+func (sg *PxfServiceGroup) validate(c *PxfCluster) error {
+	validators := []pxfServiceGroupValidator{
+		pxfServiceGroupAtLeastOnePort,
+		pxfServiceGroupValidPorts,
+		pxfServiceGroupPortsWarning,
+	}
+	// apply group-level validators
+	for _, validator := range validators {
+		if err := validator(c, sg); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// validate applies host-level validation rules
+// a pointer to a cluster needs to be passed to give proper context for the error message
+func (h *PxfHost) validate(c *PxfCluster) error {
+	// apply group-level validators
+	return pxfHostValidHostname(h, c)
+}
+
+// sortedMapValues returns a slice of values from the provided map in the order of their sorted keys
+func sortedMapValues[K cmp.Ordered, V any](collection map[K]*V) []*V {
+	keys := make([]K, 0, len(collection))
+	for key, _ := range collection {
```

Comment:
```suggestion
	for key := range collection {
```
Nit: the '_' is not needed if it's the second value returned from a for-range loop.